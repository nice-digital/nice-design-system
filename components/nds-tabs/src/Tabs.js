import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { slugify } from "@nice-digital/nds-core";
import { Tab } from "./Tab";
import "./../scss/tabs.scss";
export { Tab };

export const keyCodes = {
	enter: 13,
	space: 32,
	end: 35,
	home: 36,
	/**
	 * Code for the left key (37)
	 * @type Number
	 */
	left: 37,
	up: 38,
	right: 39,
	down: 40
};

export class Tabs extends Component {
	constructor(props) {
		super(props);

		this.state = {
			/**
			 * The 0-based index of the active tab
			 *
			 * @type number
			 */
			index: 0,
			canUseDOM: false,
			/**
			 * Whether to move focus to the active tab button
			 *
			 * @type boolean
			 */
			focusActiveTabButton: false
		};
	}

	changeTab(index, hash) {
		this.setState({
			index: index,
			focusActiveTabButton: true
		});
		this.props.changeCallback && this.props.changeCallback(hash);
	}

	getIndexOfHashThatMatchesTab(hash) {
		const tabs = this.getTabChildElements();
		const hashIndex = tabs
			.map(tab => "#" + slugify(tab.props.title))
			.indexOf(hash);
		return hashIndex === -1 ? 0 : hashIndex;
	}

	componentDidMount() {
		this.setState({
			canUseDOM: true,
			index: this.getIndexOfHashThatMatchesTab(this.props.initialHash)
		});
	}

	handleTabButtonClick(event, index) {
		const hash = event.target.dataset.hash;
		this.changeTab(index, hash);
	}

	handleTabButtonKey(e, i) {
		let newIndex = i;

		switch (e.which) {
			case keyCodes.left:
			case keyCodes.up:
				e.preventDefault();
				newIndex--;
				if (newIndex < 0) newIndex = this.getTabChildElements().length - 1;
				break;
			case keyCodes.right:
			case keyCodes.down:
				e.preventDefault();
				newIndex++;
				if (newIndex >= this.getTabChildElements().length) newIndex = 0;
				break;
			case keyCodes.home:
				newIndex = 0;
				break;
			case keyCodes.end:
				newIndex = this.getTabChildElements().length - 1;
				break;
			default:
				break;
		}

		newIndex !== i && this.changeTab(newIndex);
	}

	getTabChildElements() {
		return React.Children.toArray(this.props.children).filter(
			c => c.type === Tab
		);
	}

	render() {
		const tabs = this.getTabChildElements();

		const getTabSlug = (title, id = null) => id || slugify(title);

		const classes = classnames([
			"tabs",
			this.state.canUseDOM && "js",
			this.props.className
		]);

		const filteredProps = Object.assign({}, this.props);

		// remove from being spread on to the DOM
		delete filteredProps?.changeCallback;

		return (
			<div className={classes} {...filteredProps}>
				<ul className="tabs__list" role="tablist">
					{tabs.map((tab, i) => {
						const { title, id } = tab.props;
						const tabSlug = getTabSlug(title, id);
						const isTabActive = i === this.state.index;

						return (
							<li className="tabs__tab" key={tabSlug} role="presentation">
								<button
									className="tabs__tab-btn"
									type="button"
									role="tab"
									data-hash={tabSlug}
									id={`tab-button-${tabSlug}`}
									aria-controls={`tab-pane-${tabSlug}`}
									aria-selected={isTabActive}
									onClick={e => this.handleTabButtonClick(e, i)}
									onKeyDown={e => this.handleTabButtonKey(e, i)}
									ref={btn => {
										this.state.focusActiveTabButton &&
											isTabActive &&
											btn &&
											btn.focus();
									}}
								>
									{title}
								</button>
							</li>
						);
					})}
				</ul>
				<div className="tabs__content">
					{tabs.map((tab, i) => {
						const { children, title, id, ...rest } = tab.props;
						const tabSlug = getTabSlug(title, id);
						const isTabActive = i === this.state.index;
						return (
							<div
								className="tabs__pane"
								key={tabSlug}
								role="tabpanel"
								id={`tab-pane-${tabSlug}`}
								aria-labelledby={`tab-button-${tabSlug}`}
								aria-hidden={!isTabActive}
								{...rest}
							>
								{typeof children === "string" ? (
									<div dangerouslySetInnerHTML={{ __html: children }} />
								) : (
									children
								)}
							</div>
						);
					})}
				</div>
			</div>
		);
	}
}

Tabs.propTypes = {
	children: PropTypes.oneOfType([PropTypes.arrayOf(Tab), Tab]).isRequired,
	changeCallback: PropTypes.func,
	initialHash: PropTypes.string,
	className: PropTypes.string
};
