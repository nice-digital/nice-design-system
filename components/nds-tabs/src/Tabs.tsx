import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import { slugify } from "@nice-digital/nds-core";
import { Tab } from "./Tab";
import "./../scss/tabs.scss";
export { Tab };

export interface TabsProps {
	[prop: string]: unknown;
	children: React.ReactNode | React.ReactNode[];
	className?: string;
	onTabChange?(hash: string): void;
}

export const keyCodes = {
	enter: 13,
	space: 32,
	end: 35,
	home: 36,
	/**
	 * Code for the left key (37)
	 * @type Number
	 */
	left: 37,
	up: 38,
	right: 39,
	down: 40
};

export const Tabs: React.FC<TabsProps> = (props: TabsProps) => {
	const [index, setIndex] = useState(0);
	const [canUseDOM, setCanUseDOM] = useState(false);
	const [focusActiveTabButton, setFocusActiveTabButton] = useState(false);

	const { children, className, onTabChange, ...rest } = props;

	useEffect(() => {
		const hash = document?.location?.hash || null;
		setCanUseDOM(true);
		setIndex(getIndexOfHashThatMatchesTab(hash));
	}, []);

	const changeTab = (index: number, hash?: string) => {
		setIndex(index);
		setFocusActiveTabButton(true);
		onTabChange && hash && onTabChange(hash);
	};

	const getIndexOfHashThatMatchesTab = (hash: string | null) => {
		if (!hash) return 0;
		const tabs = getTabChildElements();
		const hashIndex = tabs
			// document.location.hash returns the # character with it
			.map((tab) => {
				const thisTab = tab as React.ReactElement;
				return "#" + slugify(thisTab.props.title);
			})
			.indexOf(hash);
		return hashIndex === -1 ? 0 : hashIndex;
	};

	const handleTabButtonClick = (
		event: React.MouseEvent<HTMLButtonElement>,
		index: number
	) => {
		const target = event.target;
		if (target instanceof HTMLButtonElement) {
			const hash = target.dataset.hash;
			changeTab(index, hash);
		}
	};

	const handleTabButtonKey = (
		e: React.KeyboardEvent<HTMLButtonElement>,
		i: number
	) => {
		let newIndex = i;

		switch (e.which) {
			case keyCodes.left:
			case keyCodes.up:
				e.preventDefault();
				newIndex--;
				if (newIndex < 0) newIndex = getTabChildElements().length - 1;
				break;
			case keyCodes.right:
			case keyCodes.down:
				e.preventDefault();
				newIndex++;
				if (newIndex >= getTabChildElements().length) newIndex = 0;
				break;
			case keyCodes.home:
				newIndex = 0;
				break;
			case keyCodes.end:
				newIndex = getTabChildElements().length - 1;
				break;
			default:
				break;
		}

		newIndex !== i && changeTab(newIndex);
	};

	const getTabChildElements = () => {
		return React.Children.toArray(children).filter((c) => {
			const childElement = c as React.ReactElement;
			return childElement.type === Tab;
		});
	};

	const tabs = getTabChildElements();

	const getTabSlug = (title: string, id = null) => id || slugify(title);

	const classes = classnames(["tabs", canUseDOM && "js", className]);

	const filteredProps = Object.assign({}, props);

	// remove from being spread on to the DOM
	delete filteredProps?.onTabChange;
	delete filteredProps?.className;

	return (
		<div className={classes} {...filteredProps}>
			<ul className="tabs__list" role="tablist">
				{tabs.map((tab, i) => {
					const tabElement = tab as React.ReactElement;
					const { title, id } = tabElement.props;
					const tabSlug = getTabSlug(title, id);
					const isTabActive = i === index;

					return (
						<li className="tabs__tab" key={tabSlug} role="presentation">
							<button
								className="tabs__tab-btn"
								type="button"
								role="tab"
								data-hash={tabSlug}
								id={`tab-button-${tabSlug}`}
								aria-controls={`tab-pane-${tabSlug}`}
								aria-selected={isTabActive}
								onClick={(e) => handleTabButtonClick(e, i)}
								onKeyDown={(e) => handleTabButtonKey(e, i)}
								ref={(btn) => {
									focusActiveTabButton && isTabActive && btn && btn.focus();
								}}
							>
								{title}
							</button>
						</li>
					);
				})}
			</ul>
			<div className="tabs__content">
				{tabs.map((tab, i) => {
					const tabElement = tab as React.ReactElement;
					const { children, title, id, ...rest } = tabElement.props;
					const tabSlug = getTabSlug(title, id);
					const isTabActive = i === index;
					return (
						<div
							className="tabs__pane"
							key={tabSlug}
							role="tabpanel"
							id={`tab-pane-${tabSlug}`}
							aria-labelledby={`tab-button-${tabSlug}`}
							aria-hidden={!isTabActive}
							{...rest}
						>
							{typeof children === "string" ? (
								<div dangerouslySetInnerHTML={{ __html: children }} />
							) : (
								children
							)}
						</div>
					);
				})}
			</div>
		</div>
	);
};
