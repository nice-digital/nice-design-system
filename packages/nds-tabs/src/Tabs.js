import React, { Component } from "react";
import PropTypes from "prop-types";
import { slugify } from "@nice-digital/nds-core/es/utils";
import { Tab } from "./Tab";
export { Tab };
import "./../scss/tabs.scss";

export const keyCodes = {
	enter: 13,
	space: 32,
	end: 35,
	home: 36,
	/**
	 * Code for the left key (37)
	 * @type Number
	 */
	left: 37,
	up: 38,
	right: 39,
	down: 40
};

export class Tabs extends Component {
	constructor(props) {
		super(props);

		this.state = {
			/**
			 * The 0-based index of the active tab
			 *
			 * @type number
			 */
			index: 0,
			canUseDOM: false,
			/**
			 * Whether to move focus to the active tab button
			 *
			 * @type boolean
			 */
			focusActiveTabButton: false
		};
	}

	componentDidMount() {
		this.setState({ canUseDOM: true });
	}

	handleTabButtonClick(index) {
		this.setState({
			index: index,
			focusActiveTabButton: true
		});
	}

	handleTabButtonKey(e, i) {
		let newIndex = i;

		switch (e.which) {
			case keyCodes.left:
			case keyCodes.up:
				newIndex--;
				if (newIndex < 0) newIndex = this.getTabChildElements().length - 1;
				break;
			case keyCodes.right:
			case keyCodes.down:
				newIndex++;
				if (newIndex >= this.getTabChildElements().length) newIndex = 0;
				break;
			case keyCodes.home:
				newIndex = 0;
				break;
			case keyCodes.end:
				newIndex = this.getTabChildElements().length - 1;
				break;
			default:
				break;
		}

		newIndex != i &&
			this.setState({
				index: newIndex,
				focusActiveTabButton: true
			});
	}

	getTabChildElements() {
		return React.Children.toArray(this.props.children).filter(
			c => c.type == Tab
		);
	}

	render() {
		const tabs = this.getTabChildElements();

		const getTabSlug = (title, id = null) => id || slugify(title);

		return (
			<div
				className={`tabs${this.state.canUseDOM ? " js" : ""}`}
				{...this.props}
			>
				<ul className="tabs__list" role="tablist">
					{tabs.map((tab, i) => {
						const { title, id, ...rest } = tab.props;
						const tabSlug = getTabSlug(title, id);
						const isTabActive = i === this.state.index;

						return (
							<li className="tabs__tab" key={tabSlug} role="presentation">
								<button
									className="tabs__tab-btn"
									type="button"
									role="tab"
									id={`tab-button-${tabSlug}`}
									aria-controls={`tab-pane-${tabSlug}`}
									aria-selected={isTabActive}
									onClick={() => this.handleTabButtonClick(i)}
									onKeyDown={e => this.handleTabButtonKey(e, i)}
									ref={btn => {
										this.state.focusActiveTabButton &&
											isTabActive &&
											btn &&
											btn.focus();
									}}
								>
									{title}
								</button>
							</li>
						);
					})}
				</ul>
				<div className="tabs__content">
					{tabs.map((tab, i) => {
						const { children, title, id, ...rest } = tab.props;
						const tabSlug = getTabSlug(title, id);
						const isTabActive = i === this.state.index;
						return (
							<div
								className="tabs__pane"
								key={tabSlug}
								role="tabpanel"
								id={`tab-pane-${tabSlug}`}
								aria-labelledby={`tab-button-${tabSlug}`}
								aria-hidden={!isTabActive}
								{...rest}
							>
								{children}
							</div>
						);
					})}
				</div>
			</div>
		);
	}
}

Tabs.propTypes = {
	children: PropTypes.oneOfType([PropTypes.arrayOf(Tab), Tab]).isRequired
};
