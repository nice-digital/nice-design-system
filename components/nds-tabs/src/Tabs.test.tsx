import React from "react";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Tabs, Tab, keyCodes } from "./Tabs";

const defaultTabs = (
	<Tabs>
		<Tab title="Tab 1">
			<p>Tab one content</p>
		</Tab>
		<Tab title="Tab 2">
			<p>Tab two content</p>
		</Tab>
		<Tab title="Tab 3">
			<p>Tab three content</p>
		</Tab>
	</Tabs>
);

function getStringMarkup() {
	return "<h2 id='stringy'><p>hello</p></h2>";
}

const stringTabs = (
	<Tabs>
		<Tab title="My Tab Title">{getStringMarkup()}</Tab>
	</Tabs>
);

describe("Tabs", () => {
	it("should match the snapshot with default content", () => {
		const wrapper = render(defaultTabs);
		expect(wrapper).toMatchSnapshot();
	});

	it("should render a string of markup if the child is not a react object", () => {
		const wrapper = render(stringTabs);
		expect(wrapper.getByRole("heading", { level: 2 })).toBeInTheDocument();
	});

	it("should activate the second tab when tab 2 button is clicked", async () => {
		const { container } = render(defaultTabs);

		expect(
			container.querySelector("#tab-pane-tab-2")?.getAttribute("aria-hidden")
		).toBe("true");

		const tabButton = container.querySelector("#tab-button-tab-2");

		userEvent.click(tabButton as Element);

		await waitFor(() => {
			expect(
				container.querySelector("#tab-pane-tab-2")?.getAttribute("aria-hidden")
			).toBe("false");
			expect(
				container
					.querySelector("#tab-button-tab-2")
					?.getAttribute("aria-selected")
			).toBe("true");
		});
	});

	it("should navigate tab content with the keyboard", async () => {
		const { container } = render(defaultTabs);

		const btn1 = container.querySelector("#tab-button-tab-1") as Element;
		const btn2 = container.querySelector("#tab-button-tab-2") as Element;
		const btn3 = container.querySelector("#tab-button-tab-3") as Element;

		userEvent.click(btn1);
		await waitFor(() => {
			expect(btn1?.getAttribute("aria-selected")).toBe("true");
		});

		// userEvent.keyboard("[ArrowRight]");
		// await waitFor(() => {
		// 	expect(btn2?.getAttribute("aria-selected")).toBe("true");
		// });

		// expect(wrapper.state().index).toEqual(1);
		// btn2.simulate("keyDown", {
		// 	which: keyCodes.right
		// });
		// expect(wrapper.state().index).toEqual(2);
		// btn3.simulate("keyDown", {
		// 	which: keyCodes.home
		// });
		// expect(wrapper.state().index).toEqual(0);
		// btn1.simulate("keyDown", {
		// 	which: keyCodes.down
		// });
		// expect(wrapper.state().index).toEqual(1);
		// btn2.simulate("keyDown", {
		// 	which: keyCodes.up
		// });
		// expect(wrapper.state().index).toEqual(0);
		// btn1.simulate("keyDown", {
		// 	which: keyCodes.end
		// });
		// expect(wrapper.state().index).toEqual(2);
	});

	it("should cascade additional props to the tabs and individual tab containers", () => {
		const { container } = render(
			<Tabs data-track={false}>
				<Tab title="Tab 1" data-track={true}>
					<p>Tab one content</p>
				</Tab>
			</Tabs>
		);
		expect(
			container.querySelector("div.tabs")?.getAttribute("data-track")
		).toBe("false");
		expect(
			container.querySelector("div.tabs__pane")?.getAttribute("data-track")
		).toBe("true");
	});

	it("should merge additional classes onto container", () => {
		const { container } = render(
			<Tabs className="mt--0">
				<Tab title="Tab 1">
					<p>Tab one content</p>
				</Tab>
			</Tabs>
		);

		expect(container.querySelector("div.tabs")).toHaveClass("mt--0");
	});
});
