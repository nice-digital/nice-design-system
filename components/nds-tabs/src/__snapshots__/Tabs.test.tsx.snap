// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should match the snapshot with default content 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="tabs js"
      >
        <ul
          class="tabs__list"
          role="tablist"
        >
          <li
            class="tabs__tab"
            role="presentation"
          >
            <button
              aria-controls="tab-pane-tab-1"
              aria-selected="true"
              class="tabs__tab-btn"
              data-hash="tab-1"
              id="tab-button-tab-1"
              role="tab"
              type="button"
            >
              Tab 1
            </button>
          </li>
          <li
            class="tabs__tab"
            role="presentation"
          >
            <button
              aria-controls="tab-pane-tab-2"
              aria-selected="false"
              class="tabs__tab-btn"
              data-hash="tab-2"
              id="tab-button-tab-2"
              role="tab"
              type="button"
            >
              Tab 2
            </button>
          </li>
          <li
            class="tabs__tab"
            role="presentation"
          >
            <button
              aria-controls="tab-pane-tab-3"
              aria-selected="false"
              class="tabs__tab-btn"
              data-hash="tab-3"
              id="tab-button-tab-3"
              role="tab"
              type="button"
            >
              Tab 3
            </button>
          </li>
        </ul>
        <div
          class="tabs__content"
        >
          <div
            aria-hidden="false"
            aria-labelledby="tab-button-tab-1"
            class="tabs__pane"
            id="tab-pane-tab-1"
            role="tabpanel"
          >
            <p>
              Tab one content
            </p>
          </div>
          <div
            aria-hidden="true"
            aria-labelledby="tab-button-tab-2"
            class="tabs__pane"
            id="tab-pane-tab-2"
            role="tabpanel"
          >
            <p>
              Tab two content
            </p>
          </div>
          <div
            aria-hidden="true"
            aria-labelledby="tab-button-tab-3"
            class="tabs__pane"
            id="tab-pane-tab-3"
            role="tabpanel"
          >
            <p>
              Tab three content
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="tabs js"
    >
      <ul
        class="tabs__list"
        role="tablist"
      >
        <li
          class="tabs__tab"
          role="presentation"
        >
          <button
            aria-controls="tab-pane-tab-1"
            aria-selected="true"
            class="tabs__tab-btn"
            data-hash="tab-1"
            id="tab-button-tab-1"
            role="tab"
            type="button"
          >
            Tab 1
          </button>
        </li>
        <li
          class="tabs__tab"
          role="presentation"
        >
          <button
            aria-controls="tab-pane-tab-2"
            aria-selected="false"
            class="tabs__tab-btn"
            data-hash="tab-2"
            id="tab-button-tab-2"
            role="tab"
            type="button"
          >
            Tab 2
          </button>
        </li>
        <li
          class="tabs__tab"
          role="presentation"
        >
          <button
            aria-controls="tab-pane-tab-3"
            aria-selected="false"
            class="tabs__tab-btn"
            data-hash="tab-3"
            id="tab-button-tab-3"
            role="tab"
            type="button"
          >
            Tab 3
          </button>
        </li>
      </ul>
      <div
        class="tabs__content"
      >
        <div
          aria-hidden="false"
          aria-labelledby="tab-button-tab-1"
          class="tabs__pane"
          id="tab-pane-tab-1"
          role="tabpanel"
        >
          <p>
            Tab one content
          </p>
        </div>
        <div
          aria-hidden="true"
          aria-labelledby="tab-button-tab-2"
          class="tabs__pane"
          id="tab-pane-tab-2"
          role="tabpanel"
        >
          <p>
            Tab two content
          </p>
        </div>
        <div
          aria-hidden="true"
          aria-labelledby="tab-button-tab-3"
          class="tabs__pane"
          id="tab-pane-tab-3"
          role="tabpanel"
        >
          <p>
            Tab three content
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
